packages/client/hmi-client/src/services/data.ts

```diff
@@ -42,69 +42,39 @@ const getXDDDictionaries = async () => {
-	const finalModels: Model[] = [];
-
-	//
-	// simulate applying filters to the model query
-	//
-	const ModelFilterAttributes = MODEL_FILTER_FIELDS;
-	if (term.length > 0) {
-		ModelFilterAttributes.forEach((modelAttr) => {
-			const resultsAsModels = allModels;
-			const items = resultsAsModels.filter((d) =>
-				(d[modelAttr as keyof Model] as string).toLowerCase().includes(term.toLowerCase())
-			);
-			finalModels.push(...items);
-		});
-
-		// if no models match keyword search considering the ModelFilterAttributes
-		// perhaps the keyword search match a concept name, so let's also search for that
-		if (conceptFacets) {
-			const matchingCuries = [] as string[];
-			Object.keys(conceptFacets.facets.concepts).forEach((curie) => {
-				const concept = conceptFacets?.facets.concepts[curie];
-				if (concept?.name?.toLowerCase() === term.toLowerCase()) {
-					matchingCuries.push(curie);
-				}
-			});
-			matchingCuries.forEach((curie) => {
-				const matchingResult = conceptFacets?.results.filter((r) => r.curie === curie);
-				const modelIDs = matchingResult?.map((mr) => mr.id);
-				modelIDs?.forEach((modelId) => {
-					const model = allModels.find((m) => m.id === modelId);
-					if (model) {
-						finalModels.push(model);
-					}
-				});
-			});
-		}
-	}
-
-	return term.length > 0 ? uniqBy(finalModels, ID) : allModels;
-};
-
-const filterDatasets = (
-	allDatasets: Dataset[],
+const filterAssets = (
+	allAssets: Model[] | Dataset[],
+	resourceType: ResourceType,
 	conceptFacets: ConceptFacets | null,
 	term: string
 ) => {
-	const finalDatasets: Dataset[] = [];
-
-	//
-	// simulate applying filters to the dataset query
-	//
-	const DatasetFilterAttributes = DATASET_FILTER_FIELDS;
 	if (term.length > 0) {
-		DatasetFilterAttributes.forEach((datasetAttr) => {
-			const resultsAsDatasets = allDatasets;
-			const items = resultsAsDatasets.filter((d) =>
-				(d[datasetAttr as keyof Dataset] as string).toLowerCase().includes(term.toLowerCase())
-			);
-			finalDatasets.push(...items);
+		// simulate applying filters
+		const AssetFilterAttributes: string[] =
+			resourceType === ResourceType.MODEL ? MODEL_FILTER_FIELDS : DATASET_FILTER_FIELDS; // maybe turn into switch case when other resource types have to go through here
+
+		let finalAssets: Model[] | Dataset[] = [];
+
+		AssetFilterAttributes.forEach((attribute) => {
+			switch (
+				resourceType // These switch case statements avoid the typecheck issues, maybe there is a better solution?
+			) {
+				case ResourceType.MODEL:
+					finalAssets = (allAssets as Model[]).filter((d) =>
+						(d[attribute as keyof Model] as string).toLowerCase().includes(term.toLowerCase())
+					);
+					break;
+				case ResourceType.DATASET:
+					finalAssets = (allAssets as Dataset[]).filter((d) =>
+						(d[attribute as keyof Dataset] as string).toLowerCase().includes(term.toLowerCase())
+					);
+					break;
```

Comment:
For situations like this, you can consider using generics which can help cut down boiler-plate code. 

In this particularly case there isn't a huge amount of savings because this generally need to be taken into account from the start and not in medias res. Nevertheless something like this is possible:

```
		const filterXYZ = <T>(array: T[], attr: string) =>
			array.filter(d => (d[attr as keyof T] as string).toLowerCase().includes('blah'));
```

Then later on 

```
  finalAssets = filterXYZ(allAssets as Model[], attribute);
```

But yeah, if you really want to reduce replication here the generic pattern needs to be applied across various functions so you don't have the `Model[] | Dataset[]` type declaration that can make wrangling things difficult.